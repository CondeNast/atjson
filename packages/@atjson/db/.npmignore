.cache
.vscode
dist/**/test
dist/**/node_modules
src
public
test
tmp
.npmrc
tsconfig.json
