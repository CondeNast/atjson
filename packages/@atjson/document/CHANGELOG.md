# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.24.5](https://github.com/CondeNast/atjson/compare/@atjson/document@0.24.4...@atjson/document@0.24.5) (2022-01-24)

**Note:** Version bump only for package @atjson/document

## [0.24.4](https://github.com/CondeNast/atjson/compare/@atjson/document@0.24.3...@atjson/document@0.24.4) (2021-06-01)

**Note:** Version bump only for package @atjson/document

## [0.24.3](https://github.com/CondeNast/atjson/compare/@atjson/document@0.24.2...@atjson/document@0.24.3) (2020-09-22)

**Note:** Version bump only for package @atjson/document

## [0.24.2](https://github.com/CondeNast/atjson/compare/@atjson/document@0.24.1...@atjson/document@0.24.2) (2020-05-20)

### Bug Fixes

- pass slice filter override in conversion docs ([172b4d9](https://github.com/CondeNast/atjson/commit/172b4d9dd4cc4c9162375e0f7135f2e3c908feda))

## [0.24.1](https://github.com/CondeNast/atjson/compare/@atjson/document@0.24.0...@atjson/document@0.24.1) (2020-05-19)

**Note:** Version bump only for package @atjson/document

# [0.24.0](https://github.com/CondeNast/atjson/compare/@atjson/document@0.23.4...@atjson/document@0.24.0) (2020-04-16)

### Features

- use hierarchical sort as default sort ([ed8ee3d](https://github.com/CondeNast/atjson/commit/ed8ee3d1f41c4afd200fdfe5eb80c4f0fad175f6))

## [0.23.4](https://github.com/CondeNast/atjson/compare/@atjson/document@0.23.3...@atjson/document@0.23.4) (2020-04-15)

### Bug Fixes

- preserve original document sort in canonical ([fe75116](https://github.com/CondeNast/atjson/commit/fe751169647e02888a207eb08747e5576bea4cc1))

## [0.23.3](https://github.com/CondeNast/atjson/compare/@atjson/document@0.23.2...@atjson/document@0.23.3) (2020-03-25)

**Note:** Version bump only for package @atjson/document

## [0.23.3-dev.0](https://github.com/CondeNast/atjson/compare/@atjson/document@0.23.2...@atjson/document@0.23.3-dev.0) (2020-03-23)

**Note:** Version bump only for package @atjson/document

## [0.23.2](https://github.com/CondeNast/atjson/compare/@atjson/document@0.23.1...@atjson/document@0.23.2) (2020-03-11)

**Note:** Version bump only for package @atjson/document

## [0.23.1](https://github.com/CondeNast/atjson/compare/@atjson/document@0.23.0...@atjson/document@0.23.1) (2020-03-02)

**Note:** Version bump only for package @atjson/document

# [0.23.0](https://github.com/CondeNast/atjson/compare/@atjson/document@0.22.11...@atjson/document@0.23.0) (2020-02-20)

### Features

- add Annotation is type narrowing function ([#423](https://github.com/CondeNast/atjson/issues/423)) ([2858a4f](https://github.com/CondeNast/atjson/commit/2858a4f707dd14d0ece5d0bc576f38363dfbe5ba))

## [0.22.11](https://github.com/CondeNast-Copilot/atjson/compare/@atjson/document@0.22.10...@atjson/document@0.22.11) (2020-02-10)

**Note:** Version bump only for package @atjson/document

## [0.22.10](https://github.com/CondeNast/atjson/compare/@atjson/document@0.22.9...@atjson/document@0.22.10) (2020-01-24)

### 🐛 Fixes

- 🐞 Fixes bug in Annotation.equals which prevents it from returning false… ([#364](https://github.com/CondeNast/atjson/issues/364))

## [0.22.9](https://github.com/CondeNast/atjson/compare/@atjson/document@0.22.8...@atjson/document@0.22.9) (2019-12-20)

### 🐛 Fixes

- 🐛 use ES2018 for modules output because nullish coalescing and optional chaining is breaking

## [0.22.8](https://github.com/CondeNast/atjson/compare/@atjson/document@0.22.7...@atjson/document@0.22.8) (2019-12-20)

**Note:** Version bump only for package @atjson/document

## [0.22.7](https://github.com/CondeNast/atjson/compare/@atjson/document@0.22.6...@atjson/document@0.22.7) (2019-12-19)

**Note:** Version bump only for package @atjson/document

## [0.22.6](https://github.com/CondeNast/atjson/compare/@atjson/document@0.22.5...@atjson/document@0.22.6) (2019-12-11)

**Note:** Version bump only for package @atjson/document

## [0.22.5](https://github.com/CondeNast/atjson/compare/@atjson/document@0.22.4...@atjson/document@0.22.5) (2019-12-04)

**Note:** Version bump only for package @atjson/document

## [0.22.4](https://github.com/CondeNast/atjson/compare/@atjson/document@0.22.3...@atjson/document@0.22.4) (2019-11-18)

**Note:** Version bump only for package @atjson/document

## [0.22.3](https://github.com/CondeNast/atjson/compare/@atjson/document@0.22.2...@atjson/document@0.22.3) (2019-11-07)

**Note:** Version bump only for package @atjson/document

## [0.22.3-dev276.0](https://github.com/CondeNast/atjson/compare/@atjson/document@0.22.2...@atjson/document@0.22.3-dev276.0) (2019-11-06)

**Note:** Version bump only for package @atjson/document

## [0.22.2](https://github.com/CondeNast/atjson/compare/@atjson/document@0.22.1...@atjson/document@0.22.2) (2019-10-08)

### 🐛 Fixes

- 🐞 override `slice` on conversion document class to return slices in the original source ([#238](https://github.com/CondeNast/atjson/issues/238))

## [0.22.1](https://github.com/CondeNast-Copilot/atjson/compare/@atjson/document@0.22.0...@atjson/document@0.22.1) (2019-10-08)

### 🐛 Fixes

- 🐞 clone annotations on conversion document ([#237](https://github.com/CondeNast-Copilot/atjson/issues/237))

## [0.22.0](https://github.com/CondeNast-Copilot/atjson/compare/@atjson/document@0.21.1...@atjson/document@0.22.0) (2019-10-04)

### ✨ New Features

- ✨ add cut method to document class ([#225](https://github.com/CondeNast-Copilot/atjson/issues/225))

## [0.21.1](https://github.com/CondeNast/atjson/compare/@atjson/document@0.21.0...@atjson/document@0.21.1) (2019-09-20)

### 🐛 Fixes

- 🐞 `clone` and `toJSON` coerced `undefined` values into `null`

## [0.21.0](https://github.com/CondeNast/atjson/compare/@atjson/document@0.20.0...@atjson/document@0.21.0) (2019-09-12)

**Note:** Version bump only for package @atjson/document

## [0.20.0](https://github.com/CondeNast-Copilot/atjson/compare/@atjson/document@0.19.0...@atjson/document@0.20.0) (2019-08-26)

### ✨ New Features

- ✨ Add equal() and canonical() on the document class

## [0.19.0](https://github.com/CondeNast-Copilot/atjson/compare/@atjson/document@0.17.0...@atjson/document@0.19.0) (2019-08-05)

**Note:** Version bump only for package @atjson/document

## [0.17.0](https://github.com/CondeNast-Copilot/atjson/compare/@atjson/document@0.15.0...@atjson/document@0.17.0) (2019-07-10)

### ✨ New Features

- ✨ Reduce magic in converting documents. ([#132](https://github.com/CondeNast-Copilot/atjson/issues/132))

### 🐛 Fixes

- 🐞 Fix references to wrong github repository ([#135](https://github.com/CondeNast-Copilot/atjson/issues/135))

### 🚨 Breaking Changes

- When converting a document, it will no longer coerce a document. You will need to do this yourself!
- You can no longer nest converters inside of another converter. Instead import `getConverterFor` from @atjson/document and use the function supplied from that.

These improvements are designed to improve the developer experience of atjson for developers. We've encountered some frustrations where:

1. Conversions wouldn't run because there are two different versions of @atjson/document installed.

   📝 If you are using atjson, update your dependencies to this version. Any version increment after this shouldn't break your converters.

2. Converters would get called, but the document was made up of only UnknownAnnotations.

   👩🏽‍🏫 You're no longer allowed to nest converters. This change was done to make converting less confusing, because it was super weird to have all your annotations be "Unknown" if you ran a converter before running the rest of your conversion code. Also, it will no longer clone the document, so this should hopefully make some conversions speedier! 🐰

3. It was unclear when convertTo would actually run any code.

   📣We've changed this so all converters are user-defined code. If there is no converter defined, it will let you know!

## [0.15.0](https://github.com/CondeNast/atjson/compare/@atjson/document@0.14.9...@atjson/document@0.15.0) (2019-04-19)

### ✨ New Features

- ✨🥃 add an interface for declaring annotation attributes ([#130](https://github.com/CondeNast/atjson/issues/130))

## [0.14.9](https://github.com/CondeNast/atjson/compare/@atjson/document@0.14.8...@atjson/document@0.14.9) (2019-04-15)

### 🐛 Fixes

- 🐛 fix slice so it includes annotations fully inside it as well as overlapping annotations

## [0.14.8](https://github.com/CondeNast/atjson/compare/@atjson/document@0.14.7...@atjson/document@0.14.8) (2019-04-15)

### 🐛 Fixes

- 🐝 fix slice so it only includes overlapping annotations from the parent document and the correct underlying text ([#125](https://github.com/CondeNast/atjson/issues/125))

## [0.14.7](https://github.com/CondeNast/atjson/compare/@atjson/document@0.14.6...@atjson/document@0.14.7) (2019-03-21)

### 🐛 Fixes

- 🐝 Fix unknown annotations passing attributes by reference and HIR optimization bug ([#122](https://github.com/CondeNast/atjson/issues/122))

## [0.14.6](https://github.com/CondeNast/atjson/compare/@atjson/document@0.14.5...@atjson/document@0.14.6) (2019-03-19)

**Note:** Version bump only for package @atjson/document

## [0.14.5](https://github.com/CondeNast/atjson/compare/@atjson/document@0.14.4...@atjson/document@0.14.5) (2019-03-18)

### 🐛 Fixes

- 🐛 reify unknown annotations when passing them into `createAnnotation` ([#120](https://github.com/CondeNast/atjson/issues/120))

## [0.14.4](https://github.com/CondeNast/atjson/compare/@atjson/document@0.14.3...@atjson/document@0.14.4) (2019-03-18)

### 🐛 Fixes

- 🚀🐛 Performance fixes ([#119](https://github.com/CondeNast/atjson/issues/119))

## [0.14.3](https://github.com/CondeNast/atjson/compare/@atjson/document@0.14.2...@atjson/document@0.14.3) (2019-03-14)

### 🐛 Fixes

- 🐝🚀 Fix performance regressions ([#118](https://github.com/CondeNast/atjson/issues/118))

## [0.14.2](https://github.com/CondeNast/atjson/compare/@atjson/document@0.14.1...@atjson/document@0.14.2) (2019-02-12)

### 🐛 Fixes

- 🐝 Allow annotation classes in document constructor ([#106](https://github.com/CondeNast/atjson/issues/106))

## [0.14.1](https://github.com/CondeNast/atjson/compare/@atjson/document@0.14.0...@atjson/document@0.14.1) (2019-01-14)

### 🐛 Fixes

- 🐞 fix parse annotation / unknown annotation nesting order

## [0.14.0](https://github.com/CondeNast/atjson/compare/@atjson/document@0.13.5...@atjson/document@0.14.0) (2019-01-09)

### ✨ New Features

- ✨ Use combined schema doc class in doc.convertTo

### 🐛 Fixes

- 🐜 Allow different vendor prefixes in attributes

## [0.13.5](https://github.com/CondeNast/atjson/compare/@atjson/document@0.13.4...@atjson/document@0.13.5) (2019-01-07)

### ✨ New Features

- ✨ Add query methods `document.match`, `namedCollection.outerJoin`, `namedCollection.forEach`, `namedCollection.reduce`, `join.outerJoin`, `join.where`, `join.forEach`

## [0.13.4](https://github.com/CondeNast/atjson/compare/@atjson/document@0.13.3...@atjson/document@0.13.4) (2018-12-11)

### 🐛 Fixes

- 🐛 fix tsc errors causing lerna to fail to build

## [0.13.3](https://github.com/CondeNast/atjson/compare/@atjson/document@0.13.2...@atjson/document@0.13.3) (2018-12-11)

**Note:** Version bump only for package @atjson/document

## [0.13.2](https://github.com/CondeNast/atjson/compare/@atjson/document@0.13.1...@atjson/document@0.13.2) (2018-12-11)

**Note:** Version bump only for package @atjson/document

## [0.13.1](https://github.com/CondeNast/atjson/compare/@atjson/document@0.13.0...@atjson/document@0.13.1) (2018-12-11)

**Note:** Version bump only for package @atjson/document

## [0.13.0](https://github.com/CondeNast/atjson/compare/@atjson/document@0.12.0...@atjson/document@0.13.0) (2018-12-11)

### ✨ New Features

- ✨ Coerce or convert to sources ([#93](https://github.com/CondeNast/atjson/issues/93))

## [0.12.0](https://github.com/CondeNast/atjson/compare/@atjson/document@0.11.0...@atjson/document@0.12.0) (2018-11-29)

### ✨ New Features

- ✨🔮 allow Annotation classes or AnnotationJSON to all Document methods ([#90](https://github.com/CondeNast/atjson/issues/90))
- ✨📡 dynamically convert between types of sources ([#88](https://github.com/CondeNast/atjson/issues/88))

## [0.11.0](https://github.com/CondeNast/atjson/compare/@atjson/document@0.9.0...@atjson/document@0.11.0) (2018-10-22)

### ✨ New Features

- ✨👑✨ Make Annotations classes instead of JS objects ([#57](https://github.com/CondeNast/atjson/issues/57))

### 🚨 Breaking Changes

- This introduces a bunch of breaking changes to AtJSON. The major change is that Annotations are now described as classes instead of a loose schema.

A summary of changes are the following:

⚠️ `@atjson/schema` is now deprecated. Instead of using the schema, instead use `@atjson/offset-annotations`, which provides a library of annotations that closely resemble those provided by `@atjson/schema`

🃏 Schemas are now defined on a document subclass. The schema is a list of annotation classes that are used to identify annotations in the document.

🛸 Any annotations not found in the schema are identified as `unknown` annotations. These annotations are available to be read and updated like any other annotation, but may have incorrect transform behavior because the text transform behavior is undefined. The default behavior should be acceptable for most cases.

🕵🏾‍♀️ `id`s are required property on Annotations. This is used to identify annotations for speedy updating for Offset

🖍 Annotations are now prefixed at rest. For the CommonMark Link annotation, it will be stored as `-commonmark-link` as the `type` and the attributes will be prefixed with `-commonmark` as well, meaning that instead of seeing `href` in the `attributes`, you will see `-commonmark-href`. When the annotation JSON is hydrated into an annotation class, prefixes are automatically removed. This prevents any collisions that may (and will) happen when converting between document types.

## [0.9.0](https://github.com/CondeNast/atjson/compare/@atjson/document@0.8.7...@atjson/document@0.9.0) (2018-10-10)

### ✨ New Features

- ✨🤠 Typed Annotation Collections / Joins! ([#85](https://github.com/CondeNast/atjson/issues/85))

## [0.8.7](https://github.com/CondeNast/atjson/compare/@atjson/document@0.8.6...@atjson/document@0.8.7) (2018-09-14)

**Note:** Version bump only for package @atjson/document

## [0.8.6](https://github.com/CondeNast/atjson/compare/@atjson/document@0.8.5...@atjson/document@0.8.6) (2018-09-07)

### ✨ New Features

- ✨ Added nested query syntax which allows for querying and altering collections of annotations.

## 0.8.5 (2018-09-04)

**Note:** Version bump only for package @atjson/document

## 0.8.4 (2018-07-25)

**Note:** Version bump only for package @atjson/document

## 0.7.16 (2018-04-27)

**Note:** Version bump only for package @atjson/document
