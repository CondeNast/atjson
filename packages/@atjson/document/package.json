{
  "name": "@atjson/document",
  "version": "0.30.3",
  "description": "Document definition for AtJSON",
  "main": "dist/commonjs/index.js",
  "module": "dist/modules/index.js",
  "types": "dist/commonjs/index.d.ts",
  "license": "Apache-2.0",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "nanoid": "5.1.5"
  }
}
