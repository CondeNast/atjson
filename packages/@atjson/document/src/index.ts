import {
  AdjacentBoundaryBehaviour,
  Annotation,
  AnnotationCollection,
  BlockAnnotation,
  Change,
  compareAnnotations,
  Deletion,
  Document,
  EdgeBehaviour,
  getConverterFor,
  InlineAnnotation,
  Insertion,
  Join,
  ObjectAnnotation,
  ParseAnnotation,
  UnknownAnnotation,
  is,
} from "./internals";

import type {
  AnnotationConstructor,
  AnnotationJSON,
  AttributesOf,
  JSON,
} from "./internals";

export {
  AdjacentBoundaryBehaviour,
  Annotation,
  AnnotationCollection,
  BlockAnnotation,
  Change,
  compareAnnotations,
  Deletion,
  EdgeBehaviour,
  getConverterFor,
  InlineAnnotation,
  Insertion,
  Join,
  ObjectAnnotation,
  ParseAnnotation,
  UnknownAnnotation,
  is,
};

export type { AnnotationConstructor, AnnotationJSON, AttributesOf, JSON };

export default Document;
