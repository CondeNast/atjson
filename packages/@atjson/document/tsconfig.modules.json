{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "outDir": "./dist/modules",
    "isolatedModules": true,
    "module": "esnext",
    "target": "ES2018"
  }
}
