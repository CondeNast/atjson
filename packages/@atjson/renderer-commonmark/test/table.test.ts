import { Block, deserialize } from "@atjson/document";
import OffsetSource, { Table } from "@atjson/offset-annotations";
import CommonmarkRenderer from "../src";

function testTable(
  tableAttributes: Omit<Block<Table>["attributes"], "dataSet">
) {
  return deserialize(
    {
      text: "￼￼ name age job notes laios 20 fighter A strange but earnest person. He really really likes monsters marcille 500 mage Difficult to get along with but very competent. Despite seeming strict and fussy, she is interested in forbidden magic... falin 18 healer She seems nice, but is actually just a people pleaser. When push comes to shove she will look out for people she loves and disregard strangers chilchuk 29 thief Looks like a child but is actually a divorced father of three. He is serious about his work and isn't interested in getting close with people",
      blocks: [
        {
          attributes: {
            schema: {
              name: "peritext",
              age: "peritext",
              job: "peritext",
              notes: "peritext",
            },
            records: [
              {
                age: {
                  jsonValue: "￼20",
                  slice: "M00000007",
                },
                job: {
                  jsonValue: "￼fighter",
                  slice: "M00000008",
                },
                name: {
                  jsonValue: "￼laios",
                  slice: "M00000006",
                },
                notes: {
                  jsonValue:
                    "￼A strange but earnest person. He ￼really ￼really likes monsters",
                  slice: "M00000009",
                },
              },
              {
                age: {
                  jsonValue: "￼500",
                  slice: "M0000000d",
                },
                job: {
                  jsonValue: "￼mage",
                  slice: "M0000000e",
                },
                name: {
                  jsonValue: "￼marcille",
                  slice: "M0000000c",
                },
                notes: {
                  jsonValue:
                    "￼Difficult to get along with but very competent. Despite seeming strict and fussy, she is interested in forbidden magic...",
                  slice: "M0000000f",
                },
              },
              {
                age: {
                  jsonValue: "￼18",
                  slice: "M00000011",
                },
                job: {
                  jsonValue: "￼healer",
                  slice: "M00000012",
                },
                name: {
                  jsonValue: "￼falin",
                  slice: "M00000010",
                },
                notes: {
                  jsonValue:
                    "￼She ￼seems nice, but is actually just a people pleaser. When push comes to shove she will look out for people she loves and disregard strangers",
                  slice: "M00000013",
                },
              },
              {
                age: {
                  jsonValue: "￼29",
                  slice: "M00000016",
                },
                job: {
                  jsonValue: "￼thief",
                  slice: "M00000017",
                },
                name: {
                  jsonValue: "￼chilchuk",
                  slice: "M00000015",
                },
                notes: {
                  jsonValue:
                    "￼Looks like a child but is actually a divorced father of three. He is serious about his work and isn't interested in getting close with people",
                  slice: "M00000018",
                },
              },
            ],
          },
          id: "dataSetId",
          parents: [],
          selfClosing: false,
          type: "data-set",
        },
        {
          attributes: { ...tableAttributes, dataSet: "dataSetId" },
          id: "B00000001",
          parents: [],
          selfClosing: false,
          type: "table",
        },
      ],
      marks: [
        {
          attributes: {
            refs: ["B00000000"],
          },
          id: "M00000000",
          range: "(3..7]",
          type: "slice",
        },
        {
          attributes: {
            refs: ["B00000000"],
          },
          id: "M00000001",
          range: "(8..11]",
          type: "slice",
        },
        {
          attributes: {
            refs: ["B00000000"],
          },
          id: "M00000002",
          range: "(12..15]",
          type: "slice",
        },
        {
          attributes: {
            url: "https://en.wikipedia.org/wiki/Delicious_in_Dungeon",
          },
          id: "M00000003",
          range: "(16..21)",
          type: "link",
        },
        {
          attributes: {},
          id: "M00000004",
          range: "(16..21]",
          type: "italic",
        },
        {
          attributes: {
            refs: ["B00000000"],
          },
          id: "M00000005",
          range: "(16..21]",
          type: "slice",
        },
        {
          attributes: {
            refs: ["B00000000"],
          },
          id: "M00000006",
          range: "(22..27]",
          type: "slice",
        },
        {
          attributes: {
            refs: ["B00000000"],
          },
          id: "M00000007",
          range: "(28..30]",
          type: "slice",
        },
        {
          attributes: {
            refs: ["B00000000"],
          },
          id: "M00000008",
          range: "(31..38]",
          type: "slice",
        },
        {
          attributes: {
            refs: ["B00000000"],
          },
          id: "M00000009",
          range: "(39..100]",
          type: "slice",
        },
        {
          attributes: {},
          id: "M0000000a",
          range: "(72..85]",
          type: "italic",
        },
        {
          attributes: {},
          id: "M0000000b",
          range: "(79..85]",
          type: "bold",
        },
        {
          attributes: {
            refs: ["B00000000"],
          },
          id: "M0000000c",
          range: "(101..109]",
          type: "slice",
        },
        {
          attributes: {
            refs: ["B00000000"],
          },
          id: "M0000000d",
          range: "(110..113]",
          type: "slice",
        },
        {
          attributes: {
            refs: ["B00000000"],
          },
          id: "M0000000e",
          range: "(114..118]",
          type: "slice",
        },
        {
          attributes: {
            refs: ["B00000000"],
          },
          id: "M0000000f",
          range: "(119..240]",
          type: "slice",
        },
        {
          attributes: {
            refs: ["B00000000"],
          },
          id: "M00000010",
          range: "(241..246]",
          type: "slice",
        },
        {
          attributes: {
            refs: ["B00000000"],
          },
          id: "M00000011",
          range: "(247..249]",
          type: "slice",
        },
        {
          attributes: {
            refs: ["B00000000"],
          },
          id: "M00000012",
          range: "(250..256]",
          type: "slice",
        },
        {
          attributes: {
            refs: ["B00000000"],
          },
          id: "M00000013",
          range: "(257..399]",
          type: "slice",
        },
        {
          attributes: {},
          id: "M00000014",
          range: "(261..266]",
          type: "italic",
        },
        {
          attributes: {
            refs: ["B00000000"],
          },
          id: "M00000015",
          range: "(400..408]",
          type: "slice",
        },
        {
          attributes: {
            refs: ["B00000000"],
          },
          id: "M00000016",
          range: "(409..411]",
          type: "slice",
        },
        {
          attributes: {
            refs: ["B00000000"],
          },
          id: "M00000017",
          range: "(412..417]",
          type: "slice",
        },
        {
          attributes: {
            refs: ["B00000000"],
          },
          id: "M00000018",
          range: "(418..559]",
          type: "slice",
        },
      ],
    },
    OffsetSource
  );
}

const JAGGED_TABLE = {
  blocks: [
    {
      attributes: {},
      id: "B00000000",
      parents: [],
      selfClosing: false,
      type: "text",
    },
    {
      attributes: {
        columns: [
          {
            name: "column 1",
          },
          {
            name: "column 2",
          },
          {
            name: "column 3",
          },
          {
            name: "column 4",
          },
        ],
        dataSet: "B00000003",
        showColumnHeaders: false,
      },
      id: "B00000001",
      parents: [],
      selfClosing: false,
      type: "table",
    },
    {
      attributes: {},
      id: "B00000002",
      parents: ["table"],
      selfClosing: false,
      type: "text",
    },
    {
      attributes: {
        records: [
          {
            "column 1": {
              jsonValue: "laios",
              slice: "M00000000",
            },
          },
          {
            "column 1": {
              jsonValue: "marcille",
              slice: "M00000001",
            },
            "column 2": {
              jsonValue: "500",
              slice: "M00000002",
            },
            "column 3": {
              jsonValue: "mage",
              slice: "M00000003",
            },
          },
          {
            "column 1": {
              jsonValue: "falin",
              slice: "M00000004",
            },
            "column 2": {
              jsonValue: "healer",
              slice: "M00000005",
            },
          },
          {
            "column 1": {
              jsonValue: "chilchuk",
              slice: "M00000006",
            },
            "column 2": {
              jsonValue: "29",
              slice: "M00000007",
            },
            "column 3": {
              jsonValue: "thief",
              slice: "M00000008",
            },
            "column 4": {
              jsonValue: `Looks like a child but is actually a divorced father of three. He is
        serious about his work and isn't interested in getting close with
        people`,
              slice: "M00000009",
            },
          },
        ],
        schema: {
          "column 1": "peritext",
          "column 2": "peritext",
          "column 3": "peritext",
          "column 4": "peritext",
        },
      },
      id: "B00000003",
      parents: ["table", "text"],
      selfClosing: false,
      type: "data-set",
    },
    {
      attributes: {},
      id: "B00000004",
      parents: ["table"],
      selfClosing: false,
      type: "text",
    },
  ],
  marks: [
    {
      attributes: {
        refs: ["B00000003"],
      },
      id: "M00000000",
      range: "(21..26]",
      type: "slice",
    },
    {
      attributes: {
        refs: ["B00000003"],
      },
      id: "M00000001",
      range: "(44..52]",
      type: "slice",
    },
    {
      attributes: {
        refs: ["B00000003"],
      },
      id: "M00000002",
      range: "(60..63]",
      type: "slice",
    },
    {
      attributes: {
        refs: ["B00000003"],
      },
      id: "M00000003",
      range: "(71..75]",
      type: "slice",
    },
    {
      attributes: {
        refs: ["B00000003"],
      },
      id: "M00000004",
      range: "(93..98]",
      type: "slice",
    },
    {
      attributes: {
        refs: ["B00000003"],
      },
      id: "M00000005",
      range: "(106..112]",
      type: "slice",
    },
    {
      attributes: {
        refs: ["B00000003"],
      },
      id: "M00000006",
      range: "(130..138]",
      type: "slice",
    },
    {
      attributes: {
        refs: ["B00000003"],
      },
      id: "M00000007",
      range: "(146..148]",
      type: "slice",
    },
    {
      attributes: {
        refs: ["B00000003"],
      },
      id: "M00000008",
      range: "(156..161]",
      type: "slice",
    },
    {
      attributes: {
        refs: ["B00000003"],
      },
      id: "M00000009",
      range: "(169..342]",
      type: "slice",
    },
  ],
  text: `￼
￼￼￼
  
     
      laios
    
     
      marcille
       500
       mage
    
     
      falin
       healer
    
     
      chilchuk
       29
       thief
       
        Looks like a child but is actually a divorced father of three. He is
        serious about his work and isn't interested in getting close with
        people
      
    
  
￼
    `,
} as any;

describe("tables", () => {
  describe("with column headings", () => {
    test("no alignment", () => {
      let document = testTable({
        columns: [
          { name: "name", slice: "M00000000" },
          { name: "age", slice: "M00000001" },
          { name: "job", slice: "M00000002" },
          { name: "notes", slice: "M00000005" },
        ],
        showColumnHeaders: true,
      });

      const markdown = CommonmarkRenderer.render(document);
      expect(markdown).toMatchSnapshot();
    });

    test("mixed alignment", () => {
      let document = testTable({
        columns: [
          { name: "name", slice: "M00000000", textAlign: "left" },
          { name: "age", slice: "M00000001", textAlign: "right" },
          { name: "job", slice: "M00000002" },
          { name: "notes", slice: "M00000005", textAlign: "center" },
        ],
        showColumnHeaders: true,
      });

      const markdown = CommonmarkRenderer.render(document);
      expect(markdown).toMatchSnapshot();
    });

    test("reordering columns", () => {
      let document = testTable({
        columns: [
          { name: "notes", slice: "M00000005" },
          { name: "job", slice: "M00000002" },
          { name: "age", slice: "M00000001" },
          { name: "name", slice: "M00000000" },
        ],
        showColumnHeaders: true,
      });

      const markdown = CommonmarkRenderer.render(document);
      expect(markdown).toMatchSnapshot();
    });

    test("omitting columns", () => {
      let document = testTable({
        columns: [
          {
            name: "name",
            slice: "M00000000",
          },
          {
            name: "job",
            slice: "M00000002",
          },
        ],
        showColumnHeaders: true,
      });

      const markdown = CommonmarkRenderer.render(document);
      expect(markdown).toMatchSnapshot();
    });
  });

  test("jagged table", () => {
    let doc = deserialize(JAGGED_TABLE, OffsetSource);

    expect(CommonmarkRenderer.render(doc)).toMatchSnapshot();
  });
});
