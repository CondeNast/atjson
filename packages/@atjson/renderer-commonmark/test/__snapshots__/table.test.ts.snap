// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tables jagged table 1`] = `
"
|          |        |       |                                                                                                                                                                               |
| -------- | ------ | ----- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| laios    |        |       |                                                                                                                                                                               |
| marcille | 500    | mage  |                                                                                                                                                                               |
| falin    | healer |       |                                                                                                                                                                               |
| chilchuk | 29     | thief | 
        Looks like a child but is actually a divorced father of three. He is
        serious about his work and isn't interested in getting close with
        people
       |

"
`;

exports[`tables with column headings mixed alignment 1`] = `
"| name     | age | job     | [*notes*](https://en.wikipedia.org/wiki/Delicious_in_Dungeon)                                                                                    |
|:-------- | ---:| ------- |:------------------------------------------------------------------------------------------------------------------------------------------------:|
| laios    | 20  | fighter | A strange but earnest person. He *really **really*** likes monsters                                                                              |
| marcille | 500 | mage    | Difficult to get along with but very competent. Despite seeming strict and fussy, she is interested in forbidden magic...                        |
| falin    | 18  | healer  | She *seems* nice, but is actually just a people pleaser. When push comes to shove she will look out for people she loves and disregard strangers |
| chilchuk | 29  | thief   | Looks like a child but is actually a divorced father of three. He is serious about his work and isn't interested in getting close with people    |

"
`;

exports[`tables with column headings no alignment 1`] = `
"| name     | age | job     | [*notes*](https://en.wikipedia.org/wiki/Delicious_in_Dungeon)                                                                                    |
| -------- | --- | ------- | ------------------------------------------------------------------------------------------------------------------------------------------------ |
| laios    | 20  | fighter | A strange but earnest person. He *really **really*** likes monsters                                                                              |
| marcille | 500 | mage    | Difficult to get along with but very competent. Despite seeming strict and fussy, she is interested in forbidden magic...                        |
| falin    | 18  | healer  | She *seems* nice, but is actually just a people pleaser. When push comes to shove she will look out for people she loves and disregard strangers |
| chilchuk | 29  | thief   | Looks like a child but is actually a divorced father of three. He is serious about his work and isn't interested in getting close with people    |

"
`;

exports[`tables with column headings omitting columns 1`] = `
"| name     | job     |
| -------- | ------- |
| laios    | fighter |
| marcille | mage    |
| falin    | healer  |
| chilchuk | thief   |

"
`;

exports[`tables with column headings reordering columns 1`] = `
"| [*notes*](https://en.wikipedia.org/wiki/Delicious_in_Dungeon)                                                                                    | job     | age | name     |
| ------------------------------------------------------------------------------------------------------------------------------------------------ | ------- | --- | -------- |
| A strange but earnest person. He *really **really*** likes monsters                                                                              | fighter | 20  | laios    |
| Difficult to get along with but very competent. Despite seeming strict and fussy, she is interested in forbidden magic...                        | mage    | 500 | marcille |
| She *seems* nice, but is actually just a people pleaser. When push comes to shove she will look out for people she loves and disregard strangers | healer  | 18  | falin    |
| Looks like a child but is actually a divorced father of three. He is serious about his work and isn't interested in getting close with people    | thief   | 29  | chilchuk |

"
`;
